rust/actyx/ax-core/src/runtime/operation/validation.rs

```diff
@@ -0,0 +1,754 @@
+    if value.as_dict().is_some() {
+        Ok(())
+    } else {
+        Err(anyhow::anyhow!("type mismatch, expected value to be a dictionary"))
+    }
+}
+
+// The idea is that this function will only check for Atoms,
+// the issue is that a Record can nest arbitrarily deep and may use non-atom types
+// ideally, I would like to _not_ use recursion since we already had the stack issue in other places;
+// so we either ignore records here and handle them somewhere else or we handle them here, some how
+fn validate_atom(value: &CborValue, ty: &TypeAtom) -> Result<(), anyhow::Error> {
+    match ty {
+        TypeAtom::Null => validate_null(value),
+        TypeAtom::Bool(refinement) => validate_bool(value, refinement),
+        TypeAtom::Number(refinement) => validate_number(value, refinement),
+        TypeAtom::Timestamp => validate_timestamp(value),
+        TypeAtom::String(refinement) => validate_string(value, refinement),
+        TypeAtom::Universal => Ok(()),
+    }
+}
+
+fn validate_record(value: &CborValue, ty: &NonEmptyVec<(Label, Type)>) -> Result<(), anyhow::Error> {
+    if let Some(value) = value.as_dict() {
+        for (label, ty) in ty.iter() {
+            match label {
+                Label::String(string) => {
+                    let cbor = CborBuilder::new().encode_str(string);
+                    let cbor = Cow::Owned(cbor);
+                    if let Some(value) = value.get(&cbor) {
+                        if let e @ Err(_) = validate(&value.decode(), ty) {
+                            return e;
+                        }
+                    } else {
+                        return Err(anyhow::anyhow!("label {} does not exist in record", string));
+                    }
+                }
+                Label::Number(number) => {
+                    let number = i128::from(*number);
+                    let number = Number::Int(number);
+                    let cbor = CborBuilder::new().encode_number(&number);
+                    let cbor = Cow::Owned(cbor);
+                    if let Some(value) = value.get(&cbor) {
+                        if let e @ Err(_) = validate(&value.decode(), ty) {
+                            return e;
+                        }
+                    } else {
+                        return Err(anyhow::anyhow!("label {:?} does not exist in record", number));
+                    }
+                }
+            }
+        }
+        Ok(())
+    } else {
+        Err(anyhow::anyhow!("expected a record (dict)"))
+    }
+}
+
+// Using logic rules and De Morgan laws to rewrite intersections and unions before we reach this method
+// would (probably) allow us to write a better (i.e. simpler) and more performant validator
+fn validate_union(value: &CborValue, ty: &(Type, Type)) -> Result<(), anyhow::Error> {
+    // Since the union operation is commutative, we always evaluate the most specific type first
+    // TODO: add tests
+    match ty.borrow() {
```

Comment:
why not just try the left one first, then the right one?

Reply:
I tried that approach first. IMO it becomes very hard to read due to nesting and this one provides an overall structure that changes little over the course of the function